File: libchess/move.py
import re
import libchess

class Move(object):
    def __init__(self, source, target, promotion = None):
        """Inits a move.

        Args:
            source: Source square.
            target: Target square.
            promotion: The piece type the pawn has been promoted to, if any.
        """
        self._source = source
        self._target = target
        self._promotion = promotion

        if promotion:
            assert target.is_backrank()
            assert promotion in "rnbq"

    def get_uci_move(self):
        """Gets an UCI move.

        Returns:
            A string like "a1a2" or "b7b8q".
        """
        promotion = ""
        if self._promotion:
            promotion = self._promotion
        return self._source.get_name() + self._target.get_name() + promotion

    def get_source(self):
        """Gets the source square.

        Returns:
            The source square.
        """
        return self._source

    def get_target(self):
        """Gets the target square.

        Returns:
            The target square.
        """
        return self._target

    def get_promotion(self):
        """Gets the promotion type.

        Returns:
            None, "r", "n", "b" or "q".
        """
        if not self._promotion:
            return None
        else:
            return self._promotion

    def __str__(self):
        return self.get_uci_move()

    def __repr__(self):
        return "Move.from_uci_move('%s')" % self.get_uci_move()

    def __eq__(self, other):
        return self.get_uci_move() == other.get_uci_move()

    def __ne__(self, other):
        return self.get_uci_move() != other.get_uci_move()

    def __hash__(self):
        return hash(self.get_uci_move())

    @staticmethod
    def from_uci_move(move):
        """Parses an UCI move like "a1a2" or "b7b8q

        Returns:
            A new move object.
        """
        uci_move_regex = re.compile("^([a-h][1-8])([a-h][1-8])([rnbq]?)$")
        match = uci_move_regex.match(move)
        return Move(libchess.Square.from_name(match.group(1)), libchess.Square.from_name(match.group(2)), match.group(3))

File: libchess/move.cpp
#include "move.h"

#include <boost/format.hpp>

namespace chess {

    Move::Move(Square source, Square target, char promotion) : m_source(source), m_target(target) {
        switch (promotion) {
            case 'r':
            case 'n':
            case 'b':
            case 'q':
                m_promotion = promotion;
                break;
            default:
                throw new std::invalid_argument("promotion");
        }
    }

    Move::Move(Square source, Square target) : m_source(source), m_target(target) {
        m_promotion = 0;
    }

    Move::Move(std::string uci) : m_source(Square(0)), m_target(Square(0)) {
        if (uci.length() == 4 || uci.length() == 5) {
            m_source = Square(uci.substr(0, 2));
            m_target = Square(uci.substr(2, 2));
            if (uci.length() == 5) {
                m_promotion = uci.at(4);
                switch (m_promotion) {
                    case 'r':
                    case 'n':
                    case 'b':
                    case 'q':
                        break;
                    default:
                        throw new std::invalid_argument("uci");
                }
            }
        }
        else {
            throw new std::invalid_argument("uci");
        }
    }

    Square Move::source() const {
        return m_source;
    }

    Square Move::target() const {
        return m_target;
    }

    char Move::promotion() const {
        return m_promotion;
    }

    std::string Move::full_promotion() const {
        switch (m_promotion) {
            case 'r':
                return "rook";
            case 'b':
                return "bishop";
            case 'n':
                return "knight";
            case 'q':
                return "queen";
        }
    }

    std::string Move::uci() const {
        if (m_promotion) {
            return m_source.name() + m_target.name() + m_promotion;
        } else {
            return m_source.name() + m_target.name();
        }
    }

    bool Move::is_promotion() const {
        return m_promotion != 0;
    }

    std::string Move::__repr__() const {
        return boost::str(boost::format("Move.from_uci('%s')") % uci());
    }

    int Move::__hash__() const {
        return m_source.__hash__() + 100 * m_target.__hash__() + 10000 * m_promotion;
    }

    bool Move::operator==(const Move& other) const {
        return m_source == other.m_source && m_target == other.m_target && m_promotion == other.m_promotion;
    }

    bool Move::operator!=(const Move& other) const {
        return m_source != other.m_source || m_target != other.m_target || m_promotion != other.m_promotion;
    }

    Move Move::from_uci(std::string uci) {
        return Move(uci);
    }

    std::ostream& operator<<(std::ostream& out, const Move& move) {
        out << move.uci();
        return out;
    }

}
